// appium-tizen-driver/build/index.js
#!/usr/bin/env node
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.commonCapConstraints = exports.startServer = exports.TizenDriver = exports.tizenCommands = exports.tizenHelpers = exports.default = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _yargs = _interopRequireDefault(require("yargs"));

var _asyncbox = require("asyncbox");

var server = _interopRequireWildcard(require("./lib/server"));

var driver = _interopRequireWildcard(require("./lib/driver"));

var helpers = _interopRequireWildcard(require("./lib/tizen-helpers"));

var commands = _interopRequireWildcard(require("./lib/commands/index"));

var caps = _interopRequireWildcard(require("./lib/desired-caps"));

const startServer = server.startServer;
exports.startServer = startServer;
const TizenDriver = driver.TizenDriver;
exports.TizenDriver = TizenDriver;
const tizenHelpers = helpers.helpers;
exports.tizenHelpers = tizenHelpers;
const tizenCommands = commands.commands;
exports.tizenCommands = tizenCommands;
const commonCapConstraints = caps.commonCapConstraints;
exports.commonCapConstraints = commonCapConstraints;
const DEFAULT_HOST = "localhost";
const DEFAULT_PORT = 4723;

function main() {
  return _main.apply(this, arguments);
}

function _main() {
  _main = (0, _asyncToGenerator2.default)(function* () {
    let port = _yargs.default.argv.port || DEFAULT_PORT;
    let host = _yargs.default.argv.host || DEFAULT_HOST;
    return yield startServer(port, host);
  });
  return _main.apply(this, arguments);
}

if (require.main === module) {
  (0, _asyncbox.asyncify)(main);
}

var _default = TizenDriver;
exports.default = _default;require('source-map-support').install();
