// appium-tizen-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TizenDriver = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _appiumBaseDriver = require("appium-base-driver");

var _desiredCaps = _interopRequireDefault(require("./desired-caps"));

var _index = _interopRequireDefault(require("./commands/index"));

var _tizenHelpers = _interopRequireDefault(require("./tizen-helpers"));

var _tizenBootstrap = _interopRequireDefault(require("./tizen-bootstrap.js"));

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumSdb = require("appium-sdb");

var _appiumSupport = require("appium-support");

const BOOTSTRAP_PORT = 8888;
const NO_PROXY = [['POST', new RegExp('^/session/[^/]+/appium')], ['GET', new RegExp('^/session/[^/]+/appium')]];

class TizenDriver extends _appiumBaseDriver.BaseDriver {
  constructor(opts = {}, shouldValidateCaps = true) {
    super(opts, shouldValidateCaps);
    this.locatorStrategies = ['id', 'accessibility id', 'class name', 'name'];
    this.desiredCapConstraints = _desiredCaps.default;
    this.jwpProxyActive = false;
    this.jwpProxyAvoid = _lodash.default.clone(NO_PROXY);
    this.settings = new _appiumBaseDriver.DeviceSettings({
      ignoreUnimportantViews: false
    });
    this.bootstrapPort = BOOTSTRAP_PORT;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _lodash.default.toPairs(_index.default)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        let _step$value = (0, _slicedToArray2.default)(_step.value, 2),
            cmd = _step$value[0],
            fn = _step$value[1];

        TizenDriver.prototype[cmd] = fn;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  createSession(caps) {
    var _this = this,
        _superprop_callCreateSession = (..._args) => super.createSession(..._args);

    return (0, _asyncToGenerator2.default)(function* () {
      try {
        let sessionId;

        var _ref = yield _superprop_callCreateSession(caps);

        var _ref2 = (0, _slicedToArray2.default)(_ref, 1);

        sessionId = _ref2[0];
        let serverDetails = {
          platform: 'LINUX',
          webStorageEnabled: false,
          takesScreenshot: false,
          javascriptEnabled: true,
          databaseEnabled: false,
          networkConnectionEnabled: false,
          locationContextEnabled: false,
          warnings: {},
          desired: _this.caps
        };
        _this.caps = Object.assign(serverDetails, _this.caps);
        let defaultOpts = {
          tmpDir: yield _appiumSupport.tempDir.staticDir(),
          fullReset: false,
          sdbPort: _appiumSdb.DEFAULT_SDB_PORT,
          tizenInstallTimeout: 50000
        };

        _lodash.default.defaults(_this.opts, defaultOpts);

        if (_this.opts.noReset === true) {
          _this.opts.fullReset = false;
        }

        if (_this.opts.fullReset === true) {
          _this.opts.noReset = false;
        }

        _this.opts.fastReset = !_this.opts.fullReset && !_this.opts.noReset;
        _this.opts.skipUninstall = _this.opts.fastReset || _this.opts.noReset;

        let _ref3 = yield _tizenHelpers.default.getDeviceInfoFromCaps(_this.opts),
            udid = _ref3.udid,
            emPort = _ref3.emPort;

        _this.opts.udid = udid;
        _this.opts.emPort = emPort;
        _this.sdb = yield _tizenHelpers.default.createSDB(_this.opts.udid, _this.opts.emPort, _this.opts.sdbPort, _this.opts.suppressKillServer);
        yield _this.startTizenSession(_this.opts);
        return [sessionId, _this.caps];
      } catch (e) {
        try {
          yield _this.deleteSession();
        } catch (ign) {}

        throw e;
      }
    })();
  }

  get appOnDevice() {
    return this.helpers.isPackageOrBundle(this.opts.appPackage);
  }

  startTizenSession() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (_this2.opts.app) {
        yield _this2.installApp(_this2.opts.app);
      }

      let isAppInstalled = yield _this2.isAppInstalled(_this2.opts.appPackage);

      if (!isAppInstalled) {
        _logger.default.errorAndThrow('Could not find to App in device.');
      }

      if (_this2.opts.appPackage) {
        let isStartedApp = yield _this2.isStartedApp();

        if (isStartedApp) {
          yield _this2.closeApp();
        }

        yield _this2.startApp({
          timeout: 20000
        });
      }

      _this2.bootstrap = new _tizenBootstrap.default(_this2.sdb, _this2.bootstrapPort, _this2.opts);
      yield _this2.bootstrap.start(_this2.opts.appPackage);

      if (_this2.opts.ignoreUnimportantViews) {
        yield _this2.settings.update({
          ignoreUnimportantViews: _this2.opts.ignoreUnimportantViews
        });
      }
    })();
  }

  checkPackagePresent() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug("Checking whether package is present on the device");

      if (!(yield _this3.sdb.shell([`app_launcher --list | grep ${_this3.opts.appPackage}`]))) {
        _logger.default.errorAndThrow(`Could not find package ${_this3.opts.appPackage} on the device`);
      }
    })();
  }

  deleteSession() {
    var _this4 = this,
        _superprop_callDeleteSession = (..._args2) => super.deleteSession(..._args2);

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug("Shutting down Tizen driver");

      yield _superprop_callDeleteSession();

      if (_this4.bootstrap) {
        yield _this4.sdb.forceStop(_this4.opts.appPackage);

        if (_this4.opts.fullReset && !_this4.opts.skipUninstall && !_this4.appOnDevice) {
          yield _this4.sdb.uninstall(_this4.opts.appPackage);
        }

        yield _this4.bootstrap.shutdown();
        _this4.bootstrap = null;
      } else {
        _logger.default.debug("Called deleteSession but bootstrap wasn't active");
      }
    })();
  }

  validateDesiredCaps(caps) {
    let res = super.validateDesiredCaps(caps);

    if (!res) {
      return res;
    }

    if (!caps.appPackage) {
      let msg = 'The desired capabilities must include an appPackage';

      _logger.default.errorAndThrow(msg);
    }
  }

  proxyActive(sessionId) {
    super.proxyActive(sessionId);
    return this.jwpProxyActive;
  }

  getProxyAvoidList(sessionId) {
    super.getProxyAvoidList(sessionId);
    return this.jwpProxyAvoid;
  }

  canProxy(sessionId) {
    super.canProxy(sessionId);
    return false;
  }

}

exports.TizenDriver = TizenDriver;
var _default = TizenDriver;
exports.default = _default;require('source-map-support').install();
