// appium-tizen-driver/build/lib/tizen-bootstrap.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.COMMAND_TYPES = exports.TizenBootstrap = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _net = _interopRequireDefault(require("net"));

var _lodash = _interopRequireDefault(require("lodash"));

var _index = _interopRequireDefault(require("./commands/index"));

var _appiumBaseDriver = require("appium-base-driver");

var _asyncbox = require("asyncbox");

var _path = _interopRequireDefault(require("path"));

var _logger = _interopRequireDefault(require("./logger"));

var _bluebird = _interopRequireDefault(require("bluebird"));

const COMMAND_TYPES = {
  ACTION: 'action',
  SHUTDOWN: 'shutdown'
};
exports.COMMAND_TYPES = COMMAND_TYPES;

class TizenBootstrap {
  constructor(sdb, systemPort = 8888, opts = {}) {
    this.appPackage;
    this.sdb = sdb;
    this.systemPort = systemPort;
    this.opts = opts;
    this.webSocket = opts.webSocket;
    this.ignoreUnexpectedShutdown = false;
    this.uiautomator = 'org.tizen.uiautomator';
    this.uiautomatorVersion = '1.0.0';
    this.isRestart = false;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _lodash.default.toPairs(_index.default)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        let _step$value = (0, _slicedToArray2.default)(_step.value, 2),
            cmd = _step$value[0],
            fn = _step$value[1];

        TizenBootstrap.prototype[cmd] = fn;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  start(appPackage) {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this.appPackage = appPackage;
      yield _this.init();
      yield _this.sdb.forwardPort(_this.systemPort, 8888);
      yield (0, _asyncbox.sleep)(6000);
      return yield _this.connectSocket();
    })();
  }

  connectSocket() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      try {
        return yield new _bluebird.default((resolve, reject) => {
          try {
            if (!_this2.socketClient) {
              _this2.socketClient = _net.default.connect(_this2.systemPort);

              _this2.socketClient.setEncoding('utf8');

              _this2.socketClient.on('error', err => {
                if (!_this2.ignoreUnexpectedShutdown) {
                  throw new Error(`Tizen bootstrap socket crashed: ${err}`);
                }
              });

              _this2.socketClient.once('connect', () => {
                _logger.default.info("Tizen bootstrap socket is now connected");

                resolve();
              });
            } else {
              _logger.default.info("SocketClient already Created");

              resolve();
            }
          } catch (err) {
            reject(err);
          }
        });
      } catch (err) {
        _logger.default.errorAndThrow(`Error occured while reconnection TizenBootstrap. Original error: ${err}`);
      }
    })();
  }

  sendCommand(type, extra = {}) {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (_this3.appPackage && type !== COMMAND_TYPES.SHUTDOWN) {
        if (_this3.isRestart) {
          yield (0, _asyncbox.sleep)(5000);
          yield _this3.restartUIAutomator();
          _this3.isRestart = false;
        }

        let isStartedApp = yield _this3.sdb.isStartedApp(_this3.appPackage);

        if (!isStartedApp) {
          yield _this3.sdb.startApp(_this3.appPackage);
          yield (0, _asyncbox.sleep)(10000);
          _this3.isRestart = false;
        }
      }

      if (!_this3.socketClient) {
        yield _this3.connectSocket();
      }

      return yield new _bluebird.default((resolve, reject) => {
        let cmd = Object.assign({
          cmd: type
        }, extra);
        let cmdJson = `${JSON.stringify(cmd)} \n`;

        _logger.default.debug(`Sending command to tizen: ${_lodash.default.truncate(cmdJson, {
          length: 1000
        }).trim()}`);

        try {
          _this3.socketClient.removeAllListeners('timeout');

          _this3.socketClient.removeAllListeners('end');

          _this3.socketClient.write(cmdJson);

          _this3.socketClient.on('data', data => {
            let streamData = '';

            _logger.default.debug(`Received command result from bootstrap : ${data}`);

            try {
              streamData = JSON.parse(streamData + data);

              _this3.socketClient.removeAllListeners('data');

              if (streamData.status === 0) {
                resolve(streamData.value);
              } else if (streamData.status === 44) {
                _this3.restartUIAutomator();

                resolve(false);
              }

              reject((0, _appiumBaseDriver.errorFromCode)(streamData.status));
            } catch (ign) {
              _logger.default.debug("Stream still not complete, waiting");

              streamData += data;
            }
          });

          _this3.socketClient.setTimeout(15000);

          _this3.socketClient.on('timeout', () => {
            _this3.socketClient.destroy();

            _this3.socketClient = null;
            _this3.isRestart = true;
            reject((0, _appiumBaseDriver.errorFromCode)(-1, "No response from Server"));
          });

          _this3.socketClient.on('end', () => {
            _this3.socketClient.destroy();

            _this3.socketClient = null;
            _this3.isRestart = true;
            reject((0, _appiumBaseDriver.errorFromCode)(-1, "Socket ended by Server"));
          });
        } catch (err) {
          reject((0, _appiumBaseDriver.errorFromCode)(-1, err));
        }
      });
    })();
  }

  sendAction(action, params = {}) {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let extra = {
        action,
        params
      };
      return yield _this4.sendCommand(COMMAND_TYPES.ACTION, extra);
    })();
  }

  shutdown() {
    var _this5 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (_this5.socketClient) {
        _this5.socketClient.end();

        _this5.socketClient.destroy();

        _this5.socketClient = null;
      }

      yield _this5.stopUIAutomator();
      yield _this5.uninstallUIAutomator();
      yield _this5.sdb.removePortForward(_this5.systemPort);
    })();
  }

  init() {
    var _this6 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let isUIAutomatorInstalled = yield _this6.isAppInstalled(_this6.uiautomator);

      if (!isUIAutomatorInstalled) {
        yield _this6.installUIAutomator();
      }

      let uiautomatorStatus = yield _this6.isStartedUIAutomator();

      if (!uiautomatorStatus) {
        yield _this6.startUIAutomator();
      } else {
        yield _this6.stopUIAutomator();
        yield (0, _asyncbox.sleep)(2000);
        yield _this6.startUIAutomator();
      }
    })();
  }

  installUIAutomator() {
    var _this7 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let arch = yield _this7.sdb.shell('uname -a');

      let rootDir = _path.default.resolve(__dirname, '..', '..');

      let tpkPath = _path.default.resolve(rootDir, 'uiautomator', `${_this7.uiautomator}-${_this7.uiautomatorVersion}`);

      if (arch.includes('i686')) {
        tpkPath += '-x86';
      } else {
        tpkPath += '-arm';
      }

      tpkPath += '.tpk';
      return yield _this7.sdb.install(tpkPath);
    })();
  }

  uninstallUIAutomator() {
    var _this8 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return yield _this8.removeApp(_this8.uiautomator);
    })();
  }

  startUIAutomator() {
    var _this9 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      yield _this9.sdb.startApp(_this9.uiautomator);
    })();
  }

  stopUIAutomator() {
    var _this10 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      yield _this10.sdb.shell(`app_launcher -t ${_this10.uiautomator}`);
    })();
  }

  restartUIAutomator() {
    var _this11 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      yield _this11.stopUIAutomator();
      yield _this11.startUIAutomator();
    })();
  }

  isStartedUIAutomator() {
    var _this12 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return yield _this12.sdb.isStartedApp(_this12.uiautomator);
    })();
  }

  set ignoreUnexpectedShutdown(ignore) {
    _logger.default.debug(`${ignore ? 'Ignoring' : 'Watching for'} bootstrap disconnect`);

    this._ignoreUnexpectedShutdown = ignore;
  }

  get ignoreUnexpectedShutdown() {
    return this._ignoreUnexpectedShutdown;
  }

}

exports.TizenBootstrap = TizenBootstrap;
var _default = TizenBootstrap;
exports.default = _default;require('source-map-support').install();
