// appium-sdb/build/lib/sdb.js
'use strict';

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _toolsIndexJs = require('./tools/index.js');

var _toolsIndexJs2 = _interopRequireDefault(_toolsIndexJs);

var DEFAULT_SDB_PORT = 26099;
var DEFAULT_OPTS = {
  sdkRoot: null,
  udid: null,
  executable: { path: "sdb", defaultArgs: [] },
  curDeviceId: null,
  emulatorPort: null,
  binaries: {},
  suppressKillServer: null,
  sdbPort: DEFAULT_SDB_PORT
};

var SDB = function SDB() {
  var opts = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  _classCallCheck(this, SDB);

  if (typeof opts.sdkRoot === "undefined") {
    opts.sdkRoot = process.env.TIZEN_HOME || '';
  }

  _Object$assign(this, opts);
  _lodash2['default'].defaultsDeep(this, _lodash2['default'].cloneDeep(DEFAULT_OPTS));

  if (opts.remoteSdbPort) {
    this.sdbPort = opts.remoteSdbPort;
  }
};

SDB.createSDB = function callee$0$0(opts) {
  var sdb;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        sdb = new SDB(opts);
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(sdb.getSdbWithCorrectSdbPath());

      case 3:
        return context$1$0.abrupt('return', sdb);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// add all the methods to the SDB prototype
var _iteratorNormalCompletion = true;
var _didIteratorError = false;
var _iteratorError = undefined;

try {
  for (var _iterator = _getIterator(_lodash2['default'].pairs(_toolsIndexJs2['default'])), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
    var _step$value = _slicedToArray(_step.value, 2);

    var fnName = _step$value[0];
    var fn = _step$value[1];

    SDB.prototype[fnName] = fn;
  }
} catch (err) {
  _didIteratorError = true;
  _iteratorError = err;
} finally {
  try {
    if (!_iteratorNormalCompletion && _iterator['return']) {
      _iterator['return']();
    }
  } finally {
    if (_didIteratorError) {
      throw _iteratorError;
    }
  }
}

exports['default'] = SDB;
exports.DEFAULT_SDB_PORT = DEFAULT_SDB_PORT;
